target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,214 @@
+{
+	switch (tag) {
+	case RB_ID_FLASH_INFO:
+		return "flash-info";
+	case RB_ID_MAC_ADDRESS_PACK:
+		return "base-mac-address";
+	case RB_ID_BOARD_PRODUCT_CODE:
+		return "board-product-code";
+	case RB_ID_BIOS_VERSION:
+		return "booter-version";
+	case RB_ID_SERIAL_NUMBER:
+		return "board-serial";
+	case RB_ID_MEMORY_SIZE:
+		return "mem-size";
+	case RB_ID_MAC_ADDRESS_COUNT:
+		return "mac-count";
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
+	case RB_ID_BOARD_IDENTIFIER:
+		return "board-identifier";
+	case RB_ID_PRODUCT_NAME:
+		return "product-name";
+	case RB_ID_DEFCONF:
+		return "defconf";
+	case RB_ID_BOARD_REVISION:
+		return "board-revision";
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_tlv_mac_read_cb(void *priv, const char *id, int index,
+                              unsigned int offset, void *buf,
+                              size_t bytes)
+{
+	eth_addr_add(buf, index);
+
+	return 0;
+}
+
+static nvmem_cell_post_process_t rb_tlv_read_cb(u16 tag, u8 *buf)
+{
+	switch (tag) {
+	case RB_ID_MAC_ADDRESS_PACK:
+		return &rb_tlv_mac_read_cb;
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_add_cells(struct device *dev, struct nvmem_device *nvmem,
+                        size_t data_len, u8 *data)
+{
+	struct nvmem_cell_info cell = {};
+	struct device_node *layout;
+	struct hard_config_tlv tlv;
+	unsigned int hdr_len = sizeof(struct hard_config_header);
+	unsigned int offset = 0;
+	int ret;
+
+	layout = of_nvmem_layout_get_container(nvmem);
+	if (!layout)
+		return -ENOENT;
+
+	while (offset < data_len) {
+		memcpy(&tlv, data + offset, sizeof(tlv));
+		if (offset + tlv.len >= data_len) {
+			dev_err(dev, "Out of bounds field (0x%x bytes at 0x%x)\n",
+				tlv.len, hdr_len + offset);
+			break;
+		}
+
+		if (tlv.tag != 0) {
+			cell.name = rb_tlv_cell_name(tlv.tag);
+			if (!cell.name)
+				goto skip;
+
+			cell.offset = hdr_len + offset + sizeof(tlv.tag) + sizeof(tlv.len);
+			/*
+			 * MikroTik stores MAC-s with length of 8 bytes,
+			 * but kernel expects it to be ETH_ALEN (6 bytes),
+			 * so we need to make sure that is the case.
+			 */
+			if (tlv.tag == RB_ID_MAC_ADDRESS_PACK)
+				cell.bytes = ETH_ALEN;
+			else
+				cell.bytes = tlv.len;
+			cell.np = of_get_child_by_name(layout, cell.name);
+			cell.read_post_process = rb_tlv_read_cb(tlv.tag, data + offset + sizeof(tlv));
+
+			ret = nvmem_add_one_cell(nvmem, &cell);
+			if (ret) {
+				of_node_put(layout);
+				return ret;
+			}
+		}
+skip:
+		offset += sizeof(tlv) + tlv.len;
+	}
+
+	of_node_put(layout);
+
+	return 0;
+}
+
+static int rb_parse_table(struct nvmem_layout *layout)
+{
+	struct nvmem_device *nvmem = layout->nvmem;
+	struct device *dev = &layout->dev;
+	struct hard_config_header hdr;
+	size_t hdr_len;
+	u8 *data;
+	int ret;
+
+	ret = nvmem_device_read(nvmem, 0, sizeof(hdr), &hdr);
+	if (ret < 0)
+		return ret;
+
+	if (memcmp(hdr.magic, RB_MAGIC_HARD, sizeof(hdr.magic))) {
```

Comment:
the magic value is CPU-endian, memcmp() will not work on both endiannesses since you're comparing byte by byte. You can simply do an u32 comparison, which is also much faster. See: https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/mtd/parsers/routerbootpart.c#L183 or https://github.com/openwrt/openwrt/blob/10a1e9b44afde93801b8be0a525f02d914f210e5/target/linux/generic/files/drivers/platform/mikrotik/rb_hardconfig.c#L715-L716

Reply:
Yeah, I moved to this as it avoids the need to care about endianess.